lib/config-files.c

```diff
@@ -0,0 +1,307 @@
+	}
+
+	return file_element;
+}
+
+#if defined(HAVE_SCANDIRAT) && defined(HAVE_OPENAT)
+
+static int filter(const struct dirent *d)
+{
+#ifdef _DIRENT_HAVE_D_TYPE
+	if (d->d_type != DT_UNKNOWN && d->d_type != DT_REG &&
+	    d->d_type != DT_LNK)
+		return 0;
+#endif
+	if (*d->d_name == '.')
+		return 0;
+
+	/* Accept this */
+	return 1;
+}
+
+static int read_dir( struct list_head *file_list,
+		     const char *project,
+		     const char *root,
+		     const char *config_name,
+		     const char *config_suffix)
+{
+	bool found = false;
+	char *dirname = NULL;
+	char *filename = NULL;
+	struct stat st;
+	int dd, nfiles, i;
+	int counter = 0;
+        struct dirent **namelist = NULL;
+	struct file_element *entry = NULL;
+
+	if (config_suffix) {
+		if (asprintf(&dirname, "%s/%s/%s.%s.d",
+			     root, project, config_name, config_suffix) < 0)
+			return -ENOMEM;
+		if (stat(dirname, &st) == 0) {
+			found = true;
+		} else {
+			free(dirname);
+			dirname = NULL;
+		}
+	}
+	if (!found) {
+		/* trying path without suffix */
+		if (asprintf(&dirname, "%s/%s/%s.d", root, project, config_name) < 0)
+			return -ENOMEM;
+		if (stat(dirname, &st) != 0) {
+			/* not found */
+			free(dirname);
+			dirname = NULL;
+		}
+	}
+
+	if (dirname==NULL)
+		return 0;
+
+	dd = open(dirname, O_RDONLY|O_CLOEXEC|O_DIRECTORY);
+	if (dd < 0) {
+		free(dirname);
+		return 0;
+	}
+
+	nfiles = scandirat(dd, ".", &namelist, filter, alphasort);
+	if (nfiles <= 0) {
+		free(dirname);
+		return 0;
+	}
+
+	for (i = 0; i < nfiles; i++) {
+		struct dirent *d = namelist[i];
+		size_t namesz = strlen(d->d_name);
+		if (strlen(config_suffix)>0 &&
+		    (!namesz || namesz < strlen(config_suffix) + 1 ||
+		     strcmp(d->d_name + (namesz - strlen(config_suffix)), config_suffix) != 0)) {
+			/* filename does not have requested suffix */
+			continue;
+		}
+
+		if (asprintf(&filename, "%s/%s", dirname, d->d_name) < 0) {
+			counter = -ENOMEM;
+			break;
+		}
+		entry = new_list_entry(filename);
+		if (entry == NULL) {
+			counter = -ENOMEM;
+			break;
+		}
+
+		list_add_tail(&entry->file_list, file_list);
+		counter++;
+	}
+
+	for (i = 0; i < nfiles; i++)
+		free(namelist[i]);
+	free(namelist);
+        free(dirname);
+	close(dd);
+	return counter;
+}
+
+#endif
+
+static void free_element(struct file_element *element)
+{
+	free(element->filename);
+}
+
+
+int config_file_list( struct list_head *file_list,
+		      const char *project,
+		      const char *etc_subdir,
+		      const char *usr_subdir,
+		      const char *config_name,
+		      const char *config_suffix)
+{
+	char *filename = NULL, *usr_basename = NULL, *etc_basename = NULL;
+	struct list_head etc_file_list;
+	struct list_head usr_file_list;
+	struct list_head *etc_entry = NULL, *usr_entry = NULL;
+	struct file_element *add_element = NULL, *usr_element = NULL, *etc_element = NULL;
+	int counter = 0;
+	
+	INIT_LIST_HEAD(file_list);
+
+	if (!config_name){
+		warnx("config_name must be a valid value");
```

Comment:
Please, without the warnx(), it writes to stderr, which is unwanted in shared libraries.